browser/misc_metrics/vertical_tab_metrics.cc

```diff
@@ -0,0 +1,189 @@
+    const TabStripSelectionChange& selection) {
+  if (!vertical_tabs_enabled_) {
+    return;
+  }
+  counts_[TabCountType::kOpen] = tab_strip_model->count();
+  TabGroupModel* group_model = tab_strip_model->group_model();
+  if (group_model != nullptr) {
+    counts_[TabCountType::kGroup] = group_model->ListTabGroups().size();
+  }
+  counts_[TabCountType::kPinned] = tab_strip_model->IndexOfFirstNonPinnedTab();
+  change_callback_.Run();
+}
+
+bool VerticalTabBrowserMetrics::IsVerticalTabsEnabled() const {
+  return vertical_tabs_enabled_;
+}
+
+size_t VerticalTabBrowserMetrics::GetTabCount(TabCountType count_type) const {
+  return counts_.at(count_type);
+}
+
+void VerticalTabBrowserMetrics::UpdateEnabledStatus() {
+  vertical_tabs_enabled_ =
+      profile_prefs_->GetBoolean(brave_tabs::kVerticalTabsEnabled);
+}
+
+VerticalTabMetrics::VerticalTabMetrics(PrefService* local_state) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    global_count_storages_[count_type] = std::make_unique<WeeklyStorage>(
+        local_state, GetStoragePrefNameForCountType(count_type));
+  }
+
+  BrowserList::GetInstance()->AddObserver(this);
+}
+
+VerticalTabMetrics::~VerticalTabMetrics() = default;
+
+void VerticalTabMetrics::RegisterPrefs(PrefRegistrySimple* registry) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    registry->RegisterListPref(GetStoragePrefNameForCountType(count_type));
+  }
+}
+
+void VerticalTabMetrics::UpdateMetrics() {
+  bool is_vertical_tabs_enabled = false;
+  for (const auto& [session_id, browser_metrics] : browser_metrics_) {
+    if (browser_metrics->IsVerticalTabsEnabled()) {
+      is_vertical_tabs_enabled = true;
+    }
+  }
+  if (!is_vertical_tabs_enabled) {
+    // Do not report if vertical tabs is not enabled on any of the
+    // active windows/profiles.
+    return;
+  }
+  // Add up tab count totals from all windows
+  base::flat_map<TabCountType, size_t> current_counts;
+  for (const auto& [session_id, browser_metrics] : browser_metrics_) {
+    for (TabCountType count_type : kAllTabCountTypes) {
+      current_counts[count_type] += browser_metrics->GetTabCount(count_type);
+    }
+  }
+  // Report histograms for each tab count type, if the
+  // particular count type is non-zero.
+  for (TabCountType count_type : kAllTabCountTypes) {
+    WeeklyStorage* storage = global_count_storages_[count_type].get();
+    storage->ReplaceTodaysValueIfGreater(current_counts[count_type]);
+    uint64_t max_value = storage->GetHighestValueInWeek();
+    if (max_value > 0) {
+      RecordMaxToHistogramBucket(count_type, max_value);
+    }
+  }
+}
+
+void VerticalTabMetrics::OnBrowserAdded(Browser* browser) {
+  if (!browser->is_type_normal()) {
+    return;
+  }
+  Profile* profile = browser->profile();
+  if (!profile || profile->IsOffTheRecord()) {
+    // Do not monitor incognito windows.
+    return;
+  }
+  PrefService* profile_prefs = profile->GetPrefs();
+  if (!profile_prefs) {
+    return;
+  }
```

Comment:
I'm wondering in which case this could happen. If this is more like to be more defensive, I'd prefer CHECK() .

Reply:
that's true, changed